fem/integrator.hpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_INTEGRATOR
+#define MFEM_INTEGRATOR
+
+#include "../config/config.hpp"
+#include "fe.hpp"
+
+namespace mfem
+{
+/** @brief This base class implements some shared functionality between
+    linear and nonlinear form integrators. */
+class Integrator
+{
+public:
+   /** @brief Create a new Integrator, optionally providing a prescribed
+       quadrature rule to use in assembly. */
+   Integrator(const IntegrationRule *ir = NULL) : IntRule(ir) {}
+
+   /** @brief Prescribe a fixed IntegrationRule to use, or set to null to let
+       the integrator choose an appropriate rule. */
+   virtual void SetIntRule(const IntegrationRule *ir) { IntRule = ir; }
+
+   /** @brief Prescribe a fixed IntegrationRule to use. */
+   void SetIntegrationRule(const IntegrationRule &ir) { SetIntRule(&ir); }
+
+   /** @brief For patchwise integration, SetNURBSPatchIntRule must be
+       called. This will override IntRule if both are non-null. */
+   void SetNURBSPatchIntRule(NURBSMeshRules *pr) { patchRules = pr; }
```

Comment:
```suggestion
   void SetNURBSPatchIntRule(NURBSMeshRules *pr)
   { patchRules = pr; if (pr) { IntRule = nullptr; } }
```
Here this one plus documentation.

Reply:
I've done this in the latest commit.